third_party/move/move-compiler-v2/src/file_format_generator/peephole_optimizer/inefficient_loads.rs

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module contains a window peephole optimizer for the Move bytecode.
+//! As with all peephole optimizers here, it assumes that the bytecode is valid.
+//!
+//! This optimizer addresses a commonly appearing pattern when involving loads.
+//!
+//! The pattern is:
+//! 1. Load a constant into the stack.
+//! 2. Store the constant into a local `u`.
+//! 3. A (possibly empty) sequence of instructions that do not involve `u`,
+//!    which we name `sequence`. Currently, the only instructions that can
+//!    involve `u` are: `CopyLoc`, `MoveLoc`, `StLoc`, `ImmBorrowLoc`,
+//!    and `MutBorrowLoc`.
+//! 4. A `MoveLoc` of `u`.
+//!
+//! This pattern can be replaced with:
+//! 1. `sequence`.
+//! 2. Load the constant into the stack.
+//!
+//! This transformation leaves the stack in the same state.
+//! The local `u` in the original code has been moved from, so later code
+//! cannot use it without a subsequent store.
+//! So, skipping the store to `u` is safe.
+
+use crate::file_format_generator::peephole_optimizer::optimizers::WindowOptimizer;
+use move_binary_format::file_format::Bytecode;
+
+/// An optimizer for inefficient loads.
+pub struct InefficientLoads;
+
+impl InefficientLoads {
+    // We need at least 3 instructions, corresponding to points 1, 2, and 4 in the pattern
+    // described in the module documentation.
+    const MIN_WINDOW_SIZE: usize = 3;
+}
+
+impl WindowOptimizer for InefficientLoads {
+    fn optimize_window(&self, window: &[Bytecode]) -> Option<(Vec<Bytecode>, usize)> {
```

Comment:
Document this function.  The `Vec` seems to be a new copy of the window which is optimized.  What is the `usize`?

Reply:
Note that this is an implementation of a trait method (`WindowOptimizer::optimize_window`), which contains the documentation:

```
    /// Given a `window` of bytecode, return a tuple containing:
    ///   1. an optimized version of a non-empty prefix of the `window`.
    ///   2. size of this prefix (should be non-zero).
    /// If `None` is returned, the `window` is not optimized.
```

I don't think it makes sense to copy paste this documentation of the trait method into every implementation: this is not the style followed either in our repo or generally in Rust or other languages I have worked with. So I am leaving this as is.